framework/src/media/voice/SoftwareEndPointDetector.cpp

```diff
@@ -50,5 +61,41 @@ bool SoftwareEndPointDetector::startEndPointDetect(int timeout)
 	return true;
 }
 
+bool SoftwareEndPointDetector::waitEndPoint(int timeout)
+{
+	int ret = -1;
+
+	if (timeout < 0) {
+		ret = sem_wait(&mSem);
+	} else {
+		struct timespec waketime;
+		clock_gettime(CLOCK_REALTIME, &waketime);
+		waketime.tv_sec += timeout;
+		ret = sem_timedwait(&mSem, &waketime);
+	}
+
+	return ret == 0 ? true : false;
+}
+
+bool SoftwareEndPointDetector::detectEndPoint(short *sample, int numSample)
+{
+	int position = 0;
+	while (position + (CONFIG_VOICE_SOFTWARE_EPD_FRAMESIZE << 1) < numSample) {
+		mPreviousVAD = mVAD;
+		mVAD = speex_preprocess_run(mState, (short *)(sample + position));	// vad : 0 (no speech) or 1 (speech)
+
+		if (mPreviousVAD == 1 && mVAD == 0) {
+			int sval;
+			medvdbg("#### EPD DETECTED!! ####\n");
+			if (sem_getvalue(&mSem, &sval) == 0 && sval < 0) {
+				sem_post(&mSem);
+			}
+			return true;
+		}
+		position += (CONFIG_VOICE_SOFTWARE_EPD_FRAMESIZE << 1);
```

Comment:
Why are you adding the doubled frame size twice here and inside while condition?

Reply:
It seems speex use 16bit data frame.
I fixed the code for better readablity.